modules/schedule_module/php/appointment.class.inc

```diff
@@ -0,0 +1,261 @@
+<?php declare(strict_types=1);
+namespace LORIS\schedule_module;
+use \Psr\Http\Message\ServerRequestInterface;
+use \Psr\Http\Message\ResponseInterface;
+/**
+ * Create the form for schedule_module/add page
+ * It includes three functions.
+ * Post method handles updating and add an appointment.
+ * Delete method handles deleting an appointment.
+ * get method handles get an appointments.*
+ * PHP Version 7
+ *
+ * @category Loris
+ * @package  Schedule_Module
+ * @author   Shen Wang <wangshen.mcin@gmail.com>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+
+class Appointment extends \NDB_Page
+{
+    public $skipTemplate = true;
+    /**
+     * Check user permissions
+     *
+     * @param \User $user The user whose access is being checked
+     *
+     * @return bool
+     */
+    function _hasAccess(\User $user) : bool
+    {
+        return $user->hasPermission('schedule_module');
+    }
+    /**
+     * Handles the duplicateCheck
+     *
+     * @param array $data the request data
+     *
+     * @return void
+     */
+    public function duplicateCheck(array $data)
+    {
+
+        $startAt = $data['AppointmentDate']." ".$data['AppointmentTime'].":00";
+        $factory = \NDB_Factory::singleton();
+        $DB      = $factory->database();
+        $duplicate_check = $DB->pselectRow(
+            "
+           SELECT 
+                * 
+            FROM 
+                appointment 
+            WHERE 
+            SessionID = :sessionId
+            And
+            AppointmentTypeID = :typeId 
+            And
+            StartsAt = :startsAt
+        ",
+            [
+                "sessionId" => $data["Session"],
+                "typeId"    => $data["AppointmentType"],
+                "startsAt"  => $startAt,
+            ]
+        );
+        if (!empty($duplicate_check)) {
+            http_response_code(400);
+            die(
+                json_encode(["error"=>"This appointment already exists."])
+            );
+        }
+    }
+    /**
+     * Handles the appointment adding process
+     *
+     * @param array $data the request data
+     *
+     * @return void
+     */
+    public function addAppointment(array $data): void
+    {
+        $this->duplicateCheck($data);
+        $factory = \NDB_Factory::singleton();
+        $DB      = $factory->database();
```

Comment:
Database handlers should be retrieved with `$this->loris->getDatabaseConnection()` when the LORIS object is available (which should always be the case here since it's extending NDB_Page which gets it passed to the constructor)